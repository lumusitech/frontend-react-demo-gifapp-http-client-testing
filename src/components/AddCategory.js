import React, { useState } from 'react'
import PropTypes from 'prop-types'

export const AddCategory = ({ setCategories }) => {

    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e) => {
        setInputValue(inputValue => e.target.value)
    };

    const handleSubmit = (e) => {
        e.preventDefault();

        if (inputValue.trim().length > 2) {
            setCategories(categories => [inputValue.trim(), ...categories]);
            setInputValue('');
        }
    };

    return (

        <form onSubmit={handleSubmit}>
            <p>{inputValue}</p>
            <input
                type="text"
                value={inputValue}
                onChange={handleInputChange}
            />
        </form>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
