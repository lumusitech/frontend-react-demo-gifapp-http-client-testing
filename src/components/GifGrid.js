import React from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs'
import PropTypes from 'prop-types'

import { GifGridItem } from './GifGridItem'

const GifGrid = ({ category }) => {

    const { data:images, loading } = useFetchGifs(category)

    return (
        <>
            <h3>{category}</h3>

            {loading && <p className="card-grid animate__animated animate__flash">Loading...</p>}

            <div className="card-grid animate__animated animate__fadeIn animate__delay-1s">
                {
                    images.map(img => (
                        <GifGridItem
                            key={img.id}
                            {...img}
                        />
                    )
                    )
                }
            </div>
        </>
    )
}

export default GifGrid

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}