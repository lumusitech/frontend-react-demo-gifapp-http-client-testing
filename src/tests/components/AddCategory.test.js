import React from 'react';
import { shallow } from 'enzyme';
import '@testing-library/jest-dom'

import { AddCategory } from '../../components/AddCategory';

describe('Testing on AddCategory.js', () => {

    // const setCategories = () => { };
    const setCategories = jest.fn(); // Mejor, crear las funciones mock
    let wrapper = shallow(<AddCategory setCategories={setCategories} />);

    beforeEach(() => {
        jest.clearAllMocks();
        wrapper = shallow(<AddCategory setCategories={setCategories} />);
    })

    test('should shows the component <AddCategory /> correctly along time', () => {

        expect(wrapper).toMatchSnapshot();
    })

    test('should change text on change', () => {
        const input = wrapper.find('input');
        const value = 'Hola mundo';
        input.simulate('change', { target: { value } });

        const p = wrapper.find('p');

        expect(p.text().trim()).toBe(value);

    })

    test('should not post info on submit', () => {
        wrapper.find('form').simulate('submit', { preventDefault() { } })

        expect(setCategories).not.toHaveBeenCalled();
    })

    test('should calls function setCategories once', () => {
        const value = 'Hola testing'
        const input = wrapper.find('input');
        input.simulate('change', { target: { value } })

        wrapper.find('form').simulate('submit', { preventDefault() { } })
        
        expect(setCategories).toHaveBeenCalled();
        expect(setCategories).toHaveBeenCalledTimes(1);
        expect(setCategories).toHaveBeenCalledWith(expect.any(Function));
        // Se verifica que después del submit, se vacíe el input
        expect(input.props().value).toBe('')
    })
})
