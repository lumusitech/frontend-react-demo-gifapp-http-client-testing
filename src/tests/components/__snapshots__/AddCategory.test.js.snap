// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing on AddCategory.js should shows the component <AddCategory /> correctly along time 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
